edb/edgeql/compiler/casts.py

```diff
@@ -540,6 +553,169 @@ def _cast_tuple(
     )
 
 
+def _cast_range(
+        ir_set: irast.Set,
+        orig_stype: s_types.Type,
+        new_stype: s_types.Type, *,
+        srcctx: Optional[parsing.ParserContext],
+        ctx: context.ContextLevel) -> irast.Set:
+
+    assert isinstance(orig_stype, s_types.Range)
+
+    direct_cast = _find_cast(orig_stype, new_stype, srcctx=srcctx, ctx=ctx)
+    if direct_cast is not None:
+        return _cast_to_ir(
+            ir_set, direct_cast, orig_stype, new_stype, ctx=ctx
+        )
+
+    if not new_stype.is_range():
+        raise errors.QueryError(
+            f'cannot cast {orig_stype.get_displayname(ctx.env.schema)!r} '
+            f'to {new_stype.get_displayname(ctx.env.schema)!r}',
+            context=srcctx)
+    assert isinstance(new_stype, s_types.Range)
+    el_type = new_stype.get_subtypes(ctx.env.schema)[0]
+    orig_el_type = orig_stype.get_subtypes(ctx.env.schema)[0]
+    ql_el_type = typegen.type_to_ql_typeref(el_type, ctx=ctx)
+
+    el_cast = _find_cast(orig_el_type, el_type, srcctx=srcctx, ctx=ctx)
+    if el_cast is None:
+        raise errors.QueryError(
+            f'cannot cast {orig_stype.get_displayname(ctx.env.schema)!r} '
+            f'to {new_stype.get_displayname(ctx.env.schema)!r}',
+            context=srcctx)
+
+    with ctx.new() as subctx:
+        subctx.anchors = subctx.anchors.copy()
+        source_path = subctx.create_anchor(ir_set, 'a')
+
+        cast = qlast.FunctionCall(
+            func=('__std__', 'range'),
+            args=[
+                qlast.TypeCast(
+                    expr=qlast.FunctionCall(
+                        func=('__std__', 'range_get_lower'),
+                        args=[source_path],
+                    ),
+                    type=ql_el_type,
+                ),
+                qlast.TypeCast(
+                    expr=qlast.FunctionCall(
+                        func=('__std__', 'range_get_upper'),
+                        args=[source_path],
+                    ),
+                    type=ql_el_type,
+                ),
+            ],
+            kwargs={
+                "inc_lower": qlast.FunctionCall(
+                    func=('__std__', 'range_is_inclusive_lower'),
+                    args=[source_path],
+                ),
+                "inc_upper": qlast.FunctionCall(
```

Comment:
We need to handle empty, also?

Reply:
> We need to handle empty, also?

Presumably we'll have the `std::range_is_empty` function as well as the `empty` argument to the range constructor that can be used here.